### PR TITLE
Implementation of BoC in C#

### PR DESCRIPTION
**Pull request by mjp41**
This is an implementation of behaviour oriented concurrency in C#. It builds on the C# task library and provides the `when` API for scheduling behaviours.

### CODE REVIEW

**Comment by mjp41** [comment]
FYI @lukecheeseman, @plietar, @MadsTorgersen

File: docs/internal/concurrency/modelimpl/StableOrder.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    private static volatile int counter = 0;
+
+    // If we get a hashcode collision, then we need to use an additional value
+    // to distinguish between the two objects. We don't want to use this from
+    // the start as it requires Interlocked operations to create it.
+    private Int64 collision = 0;
+
+    private void EnsureCollisionSet()
+    {
+        if (collision == 0)
```
Comment (plietar) [inline docs/internal/concurrency/modelimpl/StableOrder.cs:23]: Doesn't this need a CAS loop over `collision` or similar? What if two threads are scheduling a behaviour over the same cown concurrently?

Comment (mjp41) [reply docs/internal/concurrency/modelimpl/StableOrder.cs:23]: Good spot.  Thanks. I'll fix that.

Comment (TobiasWrigstad) [reply docs/internal/concurrency/modelimpl/StableOrder.cs:23]: > Alternatively, we could just use an atomic on allocation

FWIW, IMO if the goal is explaining the semantics, then this is the simpler and thus right route to take. No need to worry about collisions f.ex.

Comment (mjp41) [reply docs/internal/concurrency/modelimpl/StableOrder.cs:23]: Okay. I'll update this class to be less fun ;-)

File: docs/internal/concurrency/modelimpl/When.cs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            requests[i].StartEnqueue();
+        }
+
+        // Complete second phase of 2PL enqueuing on all cowns.
+        for (int i = 0; i < cowns.Length; i++)
+        {
+            requests[i].FinishEnqueue();
+        }
+
+        // Resolve the additional request.
```
Comment (TobiasWrigstad) [inline docs/internal/concurrency/modelimpl/When.cs:52]: The way the code and comment is written, it is a bit sneaky that this also runs (schedules) the thunk.

File: docs/internal/concurrency/modelimpl/StableOrder.cs (hunk 1 above)
Comment (TobiasWrigstad) [reply docs/internal/concurrency/modelimpl/StableOrder.cs:23]: Sorry! ;)

File: docs/internal/concurrency/modelimpl/When.cs (hunk 2 above)
Comment (TobiasWrigstad) [reply docs/internal/concurrency/modelimpl/When.cs:52]: ah sorry it doesn't!

Comment (TobiasWrigstad) [reply docs/internal/concurrency/modelimpl/When.cs:52]: OK, hope my confusion was useful. Your comment is excellent!

**Pull request merged**